src/auto_code.py

```diff
@@ -0,0 +1,196 @@
+                if rqa_key in td and not somali.DemographicCleaner.is_noise(td[rqa_key], min_length=10):
+                    is_noise = False
+            td.append_data({cls.NOISE_KEY: is_noise}, Metadata(user, Metadata.get_call_location(), time.time()))
+
+    @classmethod
+    def run_cleaners(cls, user, phone_number_uuid_table, data):
+        for plan in PipelineConfiguration.RQA_CODING_PLANS + PipelineConfiguration.FOLLOW_UP_CODING_PLANS + \
+                    PipelineConfiguration.DEMOG_CODING_PLANS:
+            for cc in plan.coding_configurations:
+                if cc.cleaner is not None:
+                    CleaningUtils.apply_cleaner_to_traced_data_iterable(user, data, plan.raw_field, cc.coded_field,
+                                                                        cc.cleaner, cc.code_scheme)
+
+        # Create a look-up table of uuids to phone numbers for all the uuids in the dataset
+        uuids = set()
+        for td in data:
+            uuids.add(td["uid"])
+        uuid_to_phone_lut = phone_number_uuid_table.uuid_to_data_batch(uuids)
+
+        # Set the operator codes for each message, using the uuid -> phone number look-up table
+        for td in data:
+            operator_clean = PhoneCleaner.clean_operator(uuid_to_phone_lut[td["uid"]])
+            if operator_clean == Codes.NOT_CODED:
+                label = CleaningUtils.make_label_from_cleaner_code(
+                    CodeSchemes.SOMALIA_OPERATOR,
+                    CodeSchemes.SOMALIA_OPERATOR.get_code_with_control_code(Codes.NOT_CODED),
+                    Metadata.get_call_location()
+                )
+            else:
+                label = CleaningUtils.make_label_from_cleaner_code(
+                    CodeSchemes.SOMALIA_OPERATOR,
+                    CodeSchemes.SOMALIA_OPERATOR.get_code_with_match_value(operator_clean),
+                    Metadata.get_call_location()
+                )
+            td.append_data({"operator_coded": label.to_dict()},
+                           Metadata(user, Metadata.get_call_location(), time.time()))
+
+    @classmethod
+    def export_coda(cls, user, data, not_noise, coda_output_dir):
+        IOUtils.ensure_dirs_exist(coda_output_dir)
+        for plan in PipelineConfiguration.RQA_CODING_PLANS:
+            if plan.coda_filename is None:
+                continue
+
+            TracedDataCodaV2IO.compute_message_ids(user, not_noise, plan.raw_field, plan.id_field)
+
+            coda_output_path = path.join(coda_output_dir, plan.coda_filename)
+            with open(coda_output_path, "w") as f:
+                TracedDataCodaV2IO.export_traced_data_iterable_to_coda_2(
+                    not_noise, plan.raw_field, plan.time_field, plan.id_field,
+                    {cc.coded_field: cc.code_scheme for cc in plan.coding_configurations},
+                    f
+                )
+
+        for plan in PipelineConfiguration.DEMOG_CODING_PLANS + PipelineConfiguration.FOLLOW_UP_CODING_PLANS:
```

Comment:
What's the reason for keeping this loop separate from the one for the RQAs?

I think the more we can keep the logic the same for all the coding plans, the less likely we are to get it wrong deep inside the pipeline source, which is what caused us problems with the q4 analysis file.

Reply:
Along that same note, what's the reason for keeping demogs and follow-ups in separate lists rather than a single 'surveys' lists as was appropriate for CSAP?

Reply:
Oh I see it's because one is for noise and the other is for normal data. Could you add a comment explaining the difference (or perhaps consider rewriting the function to take a list of coding plans and a list of data and call it twice)

Reply:
We also need ICR files for follow-ups + RQAs